File: PERSISTENCIA DE OBJETOS/Ejercicio 3/codigo/codigo.java
//3.- Sea el siguiente diagrama de clases:
//  a) Implementar el diagrama de clases.
//  b) Implementa buscarCliente(int c) a través del id.
//  c) Implementa buscarCelularCliente(int c), que devuelva los datos del cliente junto al número de celular.
import java.io.Serializable;

public class Cliente implements Serializable {
    private int id;
    private String nombre;
    private int telefono;

    public Cliente(int id, String nombre, int telefono) {
        this.id = id;
        this.nombre = nombre;
        this.telefono = telefono;
    }

    public int getId() { return id; }
    public String getNombre() { return nombre; }
    public int getTelefono() { return telefono; }

    @Override
    public String toString() {
        return "Cliente{" + "id=" + id + ", nombre='" + nombre + '\'' + ", telefono=" + telefono + '}';
    }
}

import java.io.*;

public class ArchivoCliente {
    private String nombreArchivo;

    public ArchivoCliente(String nombreArchivo) {
        this.nombreArchivo = nombreArchivo;
    }

    public void guardaCliente(Cliente c) {
        try (ObjectOutputStream oos = new ObjectOutputStream(new FileOutputStream(nombreArchivo, true))) {
            oos.writeObject(c);
            System.out.println("Cliente guardado: " + c.getNombre());
        } catch (IOException ex) {
            System.err.println("Error al guardar cliente: " + ex.getMessage());
        }
    }

    public Cliente buscarCliente(int id) {
        try (ObjectInputStream ois = new ObjectInputStream(new FileInputStream(nombreArchivo))) {
            while (true) {
                Cliente cliente = (Cliente) ois.readObject();
                if (cliente.getId() == id) {
                    return cliente;
                }
            }
        } catch (EOFException e) {
            System.out.println("Búsqueda completada");
        } catch (IOException | ClassNotFoundException ex) {
            System.err.println("Error al buscar cliente: " + ex.getMessage());
        }
        return null;
    }

    public Cliente buscarCelularCliente(int telefono) {
        try (ObjectInputStream ois = new ObjectInputStream(new FileInputStream(nombreArchivo))) {
            while (true) {
                Cliente cliente = (Cliente) ois.readObject();
                if (cliente.getTelefono() == telefono) {
                    return cliente;
                }
            }
        } catch (EOFException e) {
            System.out.println("Búsqueda completada");
        } catch (IOException | ClassNotFoundException ex) {
            System.err.println("Error al buscar cliente: " + ex.getMessage());
        }
        return null;
    }

    public static void main(String[] args) {
        ArchivoCliente archivo = new ArchivoCliente("clientes.dat");
        archivo.guardaCliente(new Cliente(1, "Ana Torres", 70123456));
        archivo.guardaCliente(new Cliente(2, "Luis Mendoza", 68754321));
        archivo.guardaCliente(new Cliente(3, "Sofía Castro", 71234567));
        Cliente encontrado = archivo.buscarCliente(2);
        System.out.println("Cliente encontrado por ID: " + encontrado);
        Cliente porTelefono = archivo.buscarCelularCliente(71234567);
        System.out.println("Cliente encontrado por teléfono: " + porTelefono);
    }
}

File: PERSISTENCIA DE OBJETOS/Ejercicio 3/codigo/codigo.py
#3.- Sea el siguiente diagrama de clases:
#   a) Implementar el diagrama de clases.
#   b) Implementa buscarCliente(int c) a través del id.
#   c) Implementa buscarCelularCliente(int c), que devuelva los datos del cliente junto al número de celular.
import pickle
class Cliente:
    def __init__(self, id_cliente, nombre, telefono):
        self.id = id_cliente
        self.nombre = nombre
        self.telefono = telefono
    
    def __str__(self):
        return f"Cliente(id={self.id}, nombre='{self.nombre}', telefono={self.telefono})"

class ArchivoCliente:
    def __init__(self, nombre_archivo):
        self.nombre_archivo = nombre_archivo
    
    def guardar_cliente(self, cliente):
        try:
            with open(self.nombre_archivo, 'rb') as f:
                clientes = pickle.load(f)
        except (FileNotFoundError, EOFError):
            clientes = []
        
        clientes.append(cliente)
        with open(self.nombre_archivo, 'wb') as f:
            pickle.dump(clientes, f)
        
        print(f"Cliente guardado: {cliente.nombre}")
    
    def buscar_cliente(self, id_cliente):
        try:
            with open(self.nombre_archivo, 'rb') as f:
                clientes = pickle.load(f)
                for cli in clientes:
                    if cli.id == id_cliente:
                        return cli
        except FileNotFoundError:
            print("Archivo no encontrado")
        except EOFError:
            print("Archivo vacío")
        return None
    
    def buscar_celular_cliente(self, telefono):
        try:
            with open(self.nombre_archivo, 'rb') as f:
                clientes = pickle.load(f)
                for cli in clientes:
                    if cli.telefono == telefono:
                        return cli
        except FileNotFoundError:
            print("Archivo no encontrado")
        except EOFError:
            print("Archivo vacío")
        return None

if __name__ == "__main__":
    archivo = ArchivoCliente("clientes.dat")
    archivo.guardar_cliente(Cliente(1, "Ana Torres", 70123456))
    archivo.guardar_cliente(Cliente(2, "Luis Mendoza", 68754321))
    archivo.guardar_cliente(Cliente(3, "Sofía Castro", 71234567))
    encontrado = archivo.buscar_cliente(2)
    print(f"Cliente encontrado por ID: {encontrado}")
    por_telefono = archivo.buscar_celular_cliente(71234567)
    print(f"Cliente encontrado por teléfono: {por_telefono}")